### DISCUSSION TITLE
Running on Pi Zero since dropping Debian packaged nodejs support

### DISCUSSION THREAD

**Discussion by deviantintegral** [Koenkk/zigbee2mqtt#13599] [General]
I have a zigbee2mqtt instance running on a Pi Zero W with Raspian 11 Bullseye.

I noticed that on `npm start` a warning is emitted that node 12 is no longer supported, which was changed in the docs at https://github.com/Koenkk/zigbee2mqtt.io/commit/8786d4590011a33d65bc876d63d713b87426b6fe

While it seems like things are running fine, I thought it would be an easy switch. I installed the node 18 repository, and re-ran `npm ci`. Even with zigbee2mqtt stopped, the Pi basically ground to a halt over a 15 minute period with significant lag, to the point where I had to reboot. It looked like mostly IO and memory contention, and the CPUs were relatively idle.

I reverted back to nodejs 12 to get the network back up.

So, it seems like it may no longer be advisable to run on Pi Zeros outside of docker images, which are precompiled. However, for a single-app device that also introduces complexity, in particular with mDNS resolution and publishing.

I think as a next step I will try building inside of a docker container from `node:18`, so at least I don't have to change apt packages to test. However, I'd love to hear from others if they have successfully built on Pi Zeros past node 12, and what I should expect for build times.

I suppose another alternative would be to use something like https://www.npmjs.com/package/pkg, which would avoid dealing with Docker and mDNS while still offering precompiled builds.

**Comment by deviantintegral** [comment]
I was able to build in a node:18 container, but only if I wasn't running zigbee2mqtt at the same time. It took around 3 minutes, which seems reasonable to me. I was then able to run the built app on the host with node 18 for a few minutes without issue.

I suppose one option would be to download the official zigbee2mqtt image, and copy out the `node_modules` folder, and run on the host. I think that would likely work given it's a debian based image.

**Comment by Koenkk** [comment]
Even though a warning is being thrown, it doesn't mean it won't work. It is not very likely that a breaking change would be introduce such that it won't work on node 12 anymore.

**Reply by deviantintegral** [reply]
Yes, that makes sense, and I think justifies only having a warning in the logs and not erroring out completely or putting something more visible in the web frontend.

Do you think it makes sense to add something to the docs along the lines of "if you're running on a low power SBC like a Pi Zero, consider using docker to avoid npm ci?" Then again, perhaps that's obvious or few users are using Pi Zeros unless others chime in with similar experiences.

**Reply by Koenkk** [reply]
What if you try with node 14 or 16? The reason it takes long is probably that no precompiled [serialport](https://github.com/serialport/node-serialport) binaries are available for node 18 triggering serialport compilation.

**Reply by deviantintegral** [reply]
I tried 16 and I had similar issues. But I will double check in case I had the version mixed up.

**Comment by EvertDekker** [comment]
There are prebuild node.js versions available for the Pi-zero with his Arm-V6 processor. No need to build yourself what most time fails on the Pi-zero.
For example install nodejs 16;
`NVM_NODEJS_ORG_MIRROR=https://unofficial-builds.nodejs.org/download/release nvm install 16`
Works great for me with Z2M 1.27.0

Source: https://dev.to/patrickweaver/installing-node-12-and-higher-on-a-raspberry-pi-zero-with-nvm-4dnj

**Reply by deviantintegral** [reply]
I think this is only required with nvm as the setup node script from nodesource worked fine for me out of the box.

**Comment by hilsonp** [comment]
@deviantintegral
Will the rpi zero w be a limiting factor for the zigbee network (with, 40+ devices) ?
Should I seek for a more powerful device or does choosing a good zigbee usb stick is the most important thing with regards to the network size ?
I still have in a box:
- a RPI 1B+
- a RIP zero w

If that is too weak to have a strong network, then I would reuse a "Up board" with the z2m docker image.

**Reply by deviantintegral** [reply]
I only ever ran that with up to 15 devices, but run-time CPU it was fine. However, the Pi Zero board actually went bad and _melted_ and that site is now using a full Raspberry Pi 4.

It's not too hard to switch your Pis. Meaning, you could start with the Zero, pair your devices, and then use `htop` to check CPU load. If you have trouble, you'll have to create a new SD card, but if you copy your `data/*` files to the new card you won't have to re-pair your devices or change anything in Home Assistant.
